### PR TITLE
Improve Wiki TOC

### PR DESCRIPTION
**Pull request by wxiaoguang**
The old code has a lot of technical debts, eg: `repo/wiki/view.tmpl` / `Iterate`

This PR improves the Wiki TOC display and improves the code.

There could be more improvements for the UI / layout, let's leave them to the future PRs.

## Before

image

## After

image

image

### CODE REVIEW

File: modules/markup/markdown/convertyaml.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ func nodeToTable(meta *yaml.Node) ast.Node {
 
 func mappingNodeToTable(meta *yaml.Node) ast.Node {
 	table := east.NewTable()
-	alignments := []east.Alignment{}
+	var alignments []east.Alignment
```
Comment (delvh) [inline modules/markup/markdown/convertyaml.go]:
```suggestion
	var alignments := make([]east.Alignment, 0, len(meta.Content)/2)
```

File: modules/markup/markdown/markdown.go [hunk 2]
```diff
@@ -73,10 +73,9 @@ func newParserContext(ctx *markup.RenderContext) parser.Context {
 	return pc
 }
 
-// actualRender renders Markdown to HTML without handling special links.
-func actualRender(ctx *markup.RenderContext, input io.Reader, output io.Writer) error {
-	once.Do(func() {
-		converter = goldmark.New(
+func SpecializedMarkdown() goldmark.Markdown {
```
Comment (delvh) [inline modules/markup/markdown/markdown.go:77]:
```suggestion
// SpecializedMarkdown sets up the Gitea specific markdown extensions
func SpecializedMarkdown() goldmark.Markdown {
```

File: routers/web/repo/wiki.go [hunk 3]
```diff
@@ -298,7 +298,15 @@ func renderViewPage(ctx *context.Context) (*git.Repository, *git.TreeEntry) {
 		ctx.Data["footerPresent"] = false
 	}
 
-	ctx.Data["toc"] = rctx.TableOfContents
+	if rctx.SidebarTocTree != nil {
+		sb := &strings.Builder{}
+		err = markdown.SpecializedMarkdown().Renderer().Render(sb, nil, rctx.SidebarTocTree)
+		if err != nil {
+			log.Error("Failed to render SidebarTocTree: %v", err)
```
Comment (delvh) [inline routers/web/repo/wiki.go]:
```suggestion
			log.Error("Failed to render wiki sidebar TOC: %v", err)
```

File: modules/markup/renderer.go [hunk 4]
```diff
@@ -63,7 +73,8 @@ type RenderContext struct {
 	GitRepo          *git.Repository
 	ShaExistCache    map[string]bool
 	cancelFn         func()
-	TableOfContents  []Header
+	SidebarTocTree   ast.Node
```
Comment (delvh) [inline modules/markup/renderer.go]: We could even think about dropping the `-Tree`:
`SidebarTOC`/`SidebarToc`

**Review by delvh — COMMENTED** [review]
I think I've understood overall what this PR refactors.

File: modules/markup/renderer.go (hunk 4 above)
Comment (wxiaoguang) [reply modules/markup/renderer.go]: Sure. 28f4b8dbdbe39817ba338769e2e5b316436909a8

I think sidebarTocNode / sidebarTocContent could be clearer than before, and `sidebarTocContent` matches the main `content` name

**Comment by silverwind** [comment]
I guess we can later re-use this toc generation mechanism later to create a toc for repo readme (via a button).

File: routers/web/base.go [hunk 5]
```diff
@@ -143,31 +143,29 @@ func Recovery(ctx goctx.Context) func(next http.Handler) http.Handler {
...
 
-					user := context.GetContextUser(req)
+					// TODO: this recovery handler is usually called without Gitea's web context, so we shouldn't touch that context too much
+					// Otherwise, the 500 page may cause new panics, eg: cache.GetContextWithData, it makes the developer&users couldn't find the original panic
+					user := context.GetContextUser(req) // almost always nil
 					if user == nil {
 						// Get user from session if logged in - do not attempt to sign-in
 						user = auth.SessionUser(sessionStore)
 					}
-					if user != nil {
```
Comment (lunny) [inline routers/web/base.go:151]: Why these lines were deleted?

Comment (wxiaoguang) [reply routers/web/base.go:151]: image

If there is `Data["SignedUser"]` , the `500` template will call `GetContextWithData` again and cause new panics.

**Review by lunny — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Comment by yardenshoham** [comment]
.

**Pull request merged**